### PR TITLE
Setup logging

### PR DESCRIPTION
**Pull request by jpnurmi**
Provides console and log file output.

Current logging:
- Subiquity socket path, server PID (level: info)
- Subiquity client requests (level: debug)
- Subiquity errors (level: error)
- Unhandled exceptions captured from Dart and Flutter (level: error)

## Configure

- Available logging levels: `debug`, `info`, `warning`, `error`

- Command-line arguments
    ```
    $ ubuntu_desktop_installer --help
    Usage: ubuntu_desktop_installer [options]

    Options:
    [...]
        --log-file=<path>         Path of the log file.
                                  Can also be specified in a LOG_FILE environment variable.
                                  (defaults to "/var/log/installer/ubuntu_desktop_installer.log")
        --log-level=<level>       Available logging levels: "debug", "info", "warning", "error".
                                  Can also be specified in a LOG_LEVEL environment variable.
                                  (defaults to "info")
    [...]
    ```

- Environment variables
  - `LOG_LEVEL=debug`
  - `LOG_FILE=/path/to/installer.log`

## Debug mode

- Default logging level: `debug`
- Default log file: `/path/to/executable/.ubuntu_desktop_installer/ubuntu_desktop_installer.log(.pid)`

## Release mode

- Default logging level: `info`
- Default log file: `/var/log/installer/ubuntu_desktop_installer.log(.pid)`

## Example

<details><summary>ubuntu_desktop_installer.log (debug)</summary>

```
2021-08-27 10:27:32.580037 INFO subiquity_server: Starting server (PID: 175437)
2021-08-27 10:27:33.615688 INFO subiquity_client: Opening socket /home/jpnurmi/Projects/canonical/ubuntu-desktop-installer/packages/ubuntu_desktop_installer/test/socket
2021-08-27 10:27:33.619045 DEBUG subiquity_client: POST http://localhost/meta/mark_configured?endpoint_names=%5B%22mirror%22%2C%22proxy%22%2C%22network%22%2C%22ssh%22%2C%22snaplist%22%2C%22timezone%22%5D
2021-08-27 10:27:33.890158 DEBUG subiquity_client: GET http://localhost/storage/has_rst
2021-08-27 10:27:33.890828 DEBUG subiquity_client: GET http://localhost/storage/has_bitlocker
2021-08-27 10:27:33.902330 DEBUG subiquity_client: GET http://localhost/keyboard
2021-08-27 10:27:35.987870 INFO subiquity_server: Starting server (PID: 175452)
2021-08-27 10:27:37.026610 INFO subiquity_client: Opening socket /home/jpnurmi/Projects/canonical/ubuntu-desktop-installer/packages/ubuntu_desktop_installer/test/socket
2021-08-27 10:27:37.032054 DEBUG subiquity_client: POST http://localhost/meta/mark_configured?endpoint_names=%5B%22mirror%22%2C%22proxy%22%2C%22network%22%2C%22ssh%22%2C%22snaplist%22%2C%22timezone%22%5D
2021-08-27 10:27:37.317206 DEBUG subiquity_client: GET http://localhost/storage/has_rst
2021-08-27 10:27:37.317713 DEBUG subiquity_client: GET http://localhost/storage/has_bitlocker
2021-08-27 10:27:37.329182 DEBUG subiquity_client: GET http://localhost/keyboard
2021-08-27 10:27:41.778536 ERROR ubuntu_desktop_installer: Unhandled exception
  _AssertionError: 'package:flutter/src/services/hardware_keyboard.dart': Failed assertion: line 441 pos 16: '_pressedKeys.containsKey(event.physicalKey)': A KeyUpEvent is dispatched, but the state shows that the physical key is not pressed. If this occurs in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the event: KeyUpEvent#9479c(physicalKey: PhysicalKeyboardKey#700e2(usbHidUsage: "0x000700e2", debugName: "Alt Left"), logicalKey: LogicalKeyboardKey#00104(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left"), character: null, timeStamp: 16:52:01.575000, synthesized)
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      HardwareKeyboard._assertEventIsRegular.<anonymous closure> (package:flutter/src/services/hardware_keyboard.dart:441:16)
#3      HardwareKeyboard._assertEventIsRegular (package:flutter/src/services/hardware_keyboard.dart:452:6)
#4      HardwareKeyboard.handleKeyEvent (package:flutter/src/services/hardware_keyboard.dart:543:5)
#5      KeyEventManager.handleRawKeyMessage (package:flutter/src/services/hardware_keyboard.dart:821:35)
#6      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:73:49)
#7      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:72:47)
#8      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:379:35)
#9      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:376:46)
#10     _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#11     _rootRun (dart:async/zone.dart:1428:13)
#12     _CustomZone.run (dart:async/zone.dart:1328:19)
#13     _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
#14     _invoke2 (dart:ui/hooks.dart:204:10)
#15     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:42:5)
#16     _Channel.push (dart:ui/channel_buffers.dart:132:31)
#17     ChannelBuffers.push (dart:ui/channel_buffers.dart:329:17)
#18     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:544:22)
#19     _dispatchPlatformMessage (dart:ui/hooks.dart:92:31)
```
</details>

Close: #261

### CODE REVIEW

**Comment by jpnurmi** [comment]
Depends on #276

File: packages/subiquity_client/lib/subiquity_client.dart [hunk 1]
```diff
@@ -2,32 +2,43 @@ library subiquity_client;
...
   Future<void> close() {
+    log.info('Closing socket ${_client.path}');
     return _client.flush().then((_) => _client.close());
   }
 
   Future<void> checkStatus(String method, StreamedResponse response) async {
     if (response.statusCode != 200) {
       final responseStr = await response.stream.bytesToString();
-      throw ("$method returned error ${response.statusCode}\n$responseStr");
+      log.error('$method returned error ${response.statusCode}\n$responseStr');
```
Comment (robert-ancell) [inline packages/subiquity_client/lib/subiquity_client.dart:30]: Should the error still be reported to the method caller? Wondering if it's safe to keep sending messages to subiquity if one has unexpectedly failed.

**Review by oSoMoN — CHANGES_REQUESTED** [review]
Looks good to me, see a few minor comments inline.
Also, there's a few places in the code where we're using `print()` that could be replaced with logging calls:

    packages/subiquity_client/lib/subiquity_server.dart:110:      print('WARNING: Error writing ${file.path} (${e.message}). '
    packages/ubuntu_desktop_installer/lib/services/disk_storage_service.dart:17:      print(r.disks);
    packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_model.dart:116:      print(e as ProcessException);
    packages/ubuntu_desktop_installer/lib/pages/installation_slides/installation_slides_model.dart:50:    print(
    packages/ubuntu_desktop_installer/lib/pages/write_changes_to_disk_page.dart:133:    print(
    packages/ubuntu_desktop_installer/lib/pages/write_changes_to_disk_page.dart:151:          print('Unexpected storage config type: ${entry['type']}');

File: packages/subiquity_client/pubspec.yaml [hunk 2]
```diff
@@ -11,8 +11,12 @@ dependencies:
   freezed_annotation: ^0.14.1
   http: ^0.13.1
   json_annotation: ^4.0.1
+  logger:
+    path: ../logger
   path: ^1.8.0
   synchronized: ^3.0.0
+  ubuntu_wizard:
+    path: ../ubuntu_wizard
```
[the remark this answers is not in the document]
Comment (jpnurmi) [reply packages/subiquity_client/pubspec.yaml]: Good catch, you're right

File: packages/logger/lib/logger.dart [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+import 'package:collection/collection.dart';
+import 'package:logging/logging.dart' as log;
+import 'package:logging_appenders/logging_appenders.dart';
+import 'package:path/path.dart' as p;
+
+/// Defines available logging levels.
+abstract class LogLevel {
+  /// All messages, including debug output.
+  static const debug = const log.Level('DEBUG', 0);
```
[the remark this answers is not in the document]
Comment (jpnurmi) [reply packages/logger/lib/logger.dart]: Looks like I never learn to run the analysis before pushing. It takes a bit too long for a commit hook too. :(

**Comment by oSoMoN** [comment]
LGTM, let's replace the handful of `print(…)` calls with logging in a follow-up PR.

**Pull request merged**